pkg/volume/vsphere_volume/vsphere_volume_util.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vsphere_volume
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	maxRetries = 10
+)
+
+// Singleton key mutex for keeping attach/detach operations for the same PD atomic
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+type VsphereDiskUtil struct{}
+
+// Attaches a disk to the current kubelet.
+// Mounts the disk to it's global path.
+func (util *VsphereDiskUtil) AttachDisk(vm *vsphereVolumeMounter, globalPDPath string) error {
+	options := []string{}
+
+	// Block execution until any pending attach/detach operations for this PD have completed
+	attachDetachMutex.LockKey(vm.volPath)
+	defer attachDetachMutex.UnlockKey(vm.volPath)
+
+	cloud, err := vm.plugin.getCloudProvider()
+	if err != nil {
+		return err
+	}
+
+	diskID, diskUUID, attachError := cloud.AttachDisk(vm.volPath, "")
+	if attachError != nil {
+		return err
+	} else if diskUUID == "" {
+		return errors.New("Disk UUID has no value")
+	}
+
+	// diskID for detach Disk
+	vm.diskID = diskID
+
+	var devicePath string
+	numTries := 0
+	for {
+		devicePath = verifyDevicePath(diskUUID)
+		// probe the attached vol so that symlink in /dev/disk/by-id is created
+		probeAttachedVolume()
+
+		_, err := os.Stat(devicePath)
+		if err == nil {
+			break
+		}
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		numTries++
+		if numTries == maxRetries {
+			return errors.New("Could not attach disk: Timeout after 60s")
+		}
+	}
+
+	notMnt, err := vm.mounter.IsLikelyNotMountPoint(globalPDPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(globalPDPath, 0750); err != nil {
+				return err
+			}
+			notMnt = true
+		} else {
+			return err
+		}
+	}
+	if notMnt {
+		err = vm.diskMounter.FormatAndMount(devicePath, globalPDPath, vm.fsType, options)
+		if err != nil {
+			os.Remove(globalPDPath)
+			return err
+		}
+		glog.V(2).Infof("Safe mount successful: %q\n", devicePath)
+	}
+	return nil
+}
+
+func verifyDevicePath(diskUUID string) string {
+	files, _ := ioutil.ReadDir("/dev/disk/by-id/")
+	for _, f := range files {
+		// TODO: should support other controllers
+		if strings.Contains(f.Name(), "scsi-") {
+			devID := f.Name()[len("scsi-"):len(f.Name())]
+			if strings.Contains(diskUUID, devID) {
```

Comment:
devID has some more characters prefixed to it compared to uuid that we get from diskUUID.